main/src/ca/uwaterloo/flix/language/ast/UnkindedType.scala

```diff
@@ -0,0 +1,601 @@
+    case object Str extends Constructor
+
+    /**
+      * A type constructor that represents the type of functions.
+      */
+    case class Arrow(arity: Int) extends Constructor
+
+    /**
+      * A type constructor that represents the type of empty records.
+      */
+    case object RecordEmpty extends Constructor
+
+    /**
+      * A type constructor that represents the type of extended records.
+      */
+    case class RecordExtend(field: Name.Field) extends Constructor
+
+    /**
+      * A type constructor that represents the type of empty schemas.
+      */
+    case object SchemaEmpty extends Constructor
+
+    /**
+      * A type constructor that represents the type of extended schemas.
+      */
+    case class SchemaExtend(pred: Name.Pred) extends Constructor
+
+    /**
+      * A type constructor that represent the type of arrays.
+      */
+    case object Array extends Constructor
+
+    /**
+      * A type constructor that represent the type of channels.
+      */
+    case object Channel extends Constructor
+
+    /**
+      * A type constructor that represent the type of lazy expressions.
+      */
+    case object Lazy extends Constructor
+
+    /**
+      * A type constructor that represent the type of tags.
+      */
+    case class Tag(sym: Symbol.EnumSym, tag: Name.Tag) extends Constructor
+
+    /**
+      * A type constructor that represent the type of enums.
+      */
+    case class Enum(sym: Symbol.EnumSym) extends Constructor
+
+    /**
+      * A type constructor that represent the type of JVM classes.
+      */
+    case class Native(clazz: Class[_]) extends Constructor
+
+    /**
+      * A type constructor that represent the type of scoped references.
+      */
+    case object ScopedRef extends Constructor
+
+    /**
+      * A type constructor that represent the type of tuples.
+      */
+    case class Tuple(l: Int) extends Constructor
+
+    /**
+      * A type constructor for relations.
+      */
+    case object Relation extends Constructor
+
+    /**
+      * A type constructor for lattices.
+      */
+    case object Lattice extends Constructor
+
+    /**
+      * A type constructor that represent the Boolean True.
+      */
+    case object True extends Constructor
+
+    /**
+      * A type constructor that represents the Boolean False.
+      */
+    case object False extends Constructor
+
+    /**
+      * A type constructor that represents the negation of an effect.
+      */
+    case object Not extends Constructor
+
+    /**
+      * A type constructor that represents the conjunction of two effects.
+      */
+    case object And extends Constructor
+
+    /**
+      * A type constructor that represents the disjunction of two effects.
+      */
+    case object Or extends Constructor
+
+    /**
+      * A type constructor that represents the type of regions.
+      */
+    case object Region extends Constructor
+
+    val Pure: Constructor = True
+    val Impure: Constructor = False
+
+  }
+}
```

Comment:
Not to go crazy, but the whole existence of UnkindedAst is because we want UnkindedType, right?

And the whole existence of UnkindedType is because a single type constructor (Var) cannot have a kind field?

Reply:
If we changed `Type` to have `UnkindedVar(...)` and then had an invariant that this does not exist after the Kinder, would everything still work?

Reply:
If this loss of type safety is too much, we could probably even use GADTs-- but that would come with its own challenges and probably not be worth it...

Reply:
Lets talk :)

Reply:
Giving up `sealed`, we could also introduce `UnkindedType` as a sub-type of Type with an extra case. We would then have to adapt certain function signatures.